be/src/storage/tablet_updates.cpp

```diff
@@ -2758,4 +2758,97 @@ Status TabletUpdates::prepare_partial_update_states(Tablet* tablet, const std::v
     return Status::OK();
 }
 
+Status TabletUpdates::get_missing_version_ranges(std::vector<int64_t>& missing_version_ranges) {
+    // usually, version ranges will have no hole or 1 hole(3 elements in ranges array)
+    missing_version_ranges.reserve(3);
+    std::lock_guard rl(_lock);
+    if (_edit_version_infos.empty()) {
+        return Status::InternalError("get_missing_version_ranges:empty edit_version_info");
+    }
+    int64_t last = _edit_version_infos.back()->version.major();
+    for (auto& itr : _pending_commits) {
+        int64_t cur = itr.first;
+        if (last + 1 < cur) {
+            missing_version_ranges.push_back(last + 1);
+            missing_version_ranges.push_back(cur - 1);
```

Comment:
can use a more structural data structure such as std::vector<std::pair<int64_t, int64_t>> since that begin and end is a pair.